crates/torii/core/src/engine.rs

```diff
@@ -119,67 +119,66 @@ impl<'db, P: Provider + Sync> Engine<'db, P> {
         Ok(latest_block_number)
     }
 
-    pub async fn sync_range(&mut self, mut from: u64, to: u64) -> Result<()> {
+    pub async fn sync_range(&mut self, from: u64, to: u64) -> Result<()> {
         // Process all blocks from current to latest.
-        while from <= to {
-            let block_with_txs = match self.provider.get_block_with_txs(BlockId::Number(from)).await
-            {
-                Ok(block_with_txs) => block_with_txs,
-                Err(e) => {
-                    error!("getting block: {}", e);
-                    continue;
-                }
-            };
+        let get_events = |token: Option<String>| {
+            self.provider.get_events(
+                EventFilter {
+                    from_block: Some(BlockId::Number(from)),
+                    to_block: Some(BlockId::Number(to)),
+                    address: Some(self.world.address),
+                    keys: None,
+                },
+                token,
+                1000,
+            )
+        };
 
-            // send the current block number
-            if let Some(ref block_tx) = self.block_tx {
-                block_tx.send(from).await.expect("failed to send block number to gRPC server");
-            }
+        // handle next events pages
+        let mut events_pages = vec![];
 
-            match self.process(block_with_txs).await {
-                Ok(_) => {
-                    self.db.set_head(from);
-                    self.db.execute().await?;
-                    from += 1;
-                }
-                Err(e) => {
-                    error!("processing block: {}", e);
-                    continue;
-                }
+        events_pages.push(get_events(None).await?);
+        while let Some(token) = &events_pages.last().unwrap().continuation_token {
+            events_pages.push(get_events(Some(token.clone())).await?);
```

Comment:
I'm wondering if using `extend` instead of a 2d array can be beneficial for CPU cache usage. We may benchmark that in the future as we will have more and more events to handled.